plenum/server/consensus/ordering_service_msg_validator.py

```diff
@@ -0,0 +1,94 @@
+from plenum.common.messages.node_messages import PrePrepare, Commit, Prepare, NewView
+from plenum.common.stashing_router import DISCARD
+from plenum.common.types import f
+from plenum.common.util import compare_3PC_keys
+from plenum.server.consensus.consensus_shared_data import ConsensusSharedData
+from plenum.server.replica_validator_enums import PROCESS, ALREADY_ORDERED, STASH_WIATING_NEW_VIEW, \
+    WAITING_FOR_NEW_VIEW, STASH_WATERMARKS, OUTSIDE_WATERMARKS, OLD_VIEW, STASH_VIEW, FUTURE_VIEW, STASH_CATCH_UP, \
+    CATCHING_UP
+
+
+class OrderingServiceMsgValidator:
+
+    def __init__(self, data: ConsensusSharedData):
+        self._data = data
+
+    def validate_pre_prepare(self, msg: PrePrepare):
+        # we discard already ordered PrePrepares
+        # check for discarded first of all
+        if self._has_already_ordered(msg):
+            return DISCARD, ALREADY_ORDERED
+
+        return self._validate_3pc(msg)
+
+    def validate_prepare(self, msg: Prepare):
+        # process Prepares that have been already ordered
+        # to re-order batches from NEW_VIEW msg
+        # prepares below low watermark have been discarded anyway
+        return self._validate_3pc(msg)
+
+    def validate_commit(self, msg: Commit):
+        # process Prepares that have been already ordered
+        # to re-order batches from NEW_VIEW msg
+        # prepares below low watermark have been discarded anyway
+        return self._validate_3pc(msg)
+
+    def validate_new_view(self, msg: NewView):
+        # View Change service has already validated NewView
+        # so basic validation here is sufficient
+        return self._validate_base(msg)
+
+    def _validate_3pc(self, msg):
+        pp_seq_no = getattr(msg, f.PP_SEQ_NO.nm, None)
+
+        # DISCARD CHECKS first
+
+        # Check if below lower watermark (meaning it's already ordered)
+        if pp_seq_no <= self._data.low_watermark:
+            return DISCARD, ALREADY_ORDERED
+
+        # Default checks next
+        res, reason = self._validate_base(msg)
+        if res != PROCESS:
+            return res, reason
+
+        # STASH CHECKS finally
+
+        # Check if waiting for new view
+        if self._data.waiting_for_new_view:
+            return STASH_WIATING_NEW_VIEW, WAITING_FOR_NEW_VIEW
+
+        # Check if above high watermarks
+        if pp_seq_no is not None and pp_seq_no > self._data.high_watermark:
+            return STASH_WATERMARKS, OUTSIDE_WATERMARKS
+
+        # PROCESS
+        return PROCESS, None
+
+    def _validate_base(self, msg):
+        view_no = getattr(msg, f.VIEW_NO.nm, None)
+
+        # DISCARD CHECKS
+
+        # Check if from old view
+        if view_no < self._data.view_no:
+            return DISCARD, OLD_VIEW
```

Comment:
Continuing comment on `original_view_no` - without it this could be done like
```python
    if view_no < self._data.view_no and batch_id not in self._data.preprepared:
        return DISCARD, OLD_VIEW
```

Reply:
I'm not sure this will be correct. I think we should not process any Prepares/Commits (for batches from NewView) in next view that were created for the corresponding PrePrepare in the previous view. 
`ppSeqNo` will be the same for both PrePrepares, only viewNo is different.

Reply:
Yes, this is clearly incorrect. What I really meant was
```python
if view_no < self._data.view_no 
    and pp_seq_no not in {batch_id.pp_seq_no for batch_id in self._data.preprepared}:
    return DISCARD, OLD_VIEW
```